### ISSUE TITLE
Infinite recursion on category edit page

### ISSUE THREAD

**Issue by yutv** [Smile-SA/elasticsuite#1629]
The infinite recursion is caused in the in the Smile\ElasticsuiteVirtualCategory\Model\Rule::[getCategorySearchQuery](https://github.com/Smile-SA/elasticsuite/blob/2.8.3/src/module-elasticsuite-virtual-category/Model/Rule.php#L139) method.

### Preconditions

Magento Version : 2.3.3

ElasticSuite Version : 2.8.3

Environment : Production

Third party modules : all custom modules (app/code), theme (app/design) and other 3-rd party modules from composer.json are removed but the issue still remains.

### Steps to reproduce

Unfortunately it is not easy to provide an exact test case because it somehow relates on a specific category tree. There is a lot of virtual categories on the website, e.g.:

    Default Category
    - Some Category
    - Virtual Category 1
    - Virtual Category 2 (it has about 180 virtual subcategories)
    - Virtual Category 3
     ... etc

The Default Category is set as root for Virtual Categories.

When you try to open the "Virtual Category 2" edit page the website runs into infinite recursion (CPU is loaded on 100% and doesn't stop, you have to reload php-fpm).

### Expected result

1. category edit page should be loaded

### Actual result

1. **504 Gateway Time-out** error from nginx
2. 100% CPU load (I reloaded php-fpm manually after a couple of minutes)

### Some debug results

It looks like the issue in the Smile\ElasticsuiteVirtualCategory\Model\Rule::[getCategorySearchQuery](https://github.com/Smile-SA/elasticsuite/blob/2.8.3/src/module-elasticsuite-virtual-category/Model/Rule.php#L139) method.

I put the following code in the line 149 and got the following log https://pastebin.com/27XAMsd2. 

            if ($handle = fopen('/tmp/test.log', 'a+')) {
                $message = "category: id={$category->getId()}, path={$category->getData('path')}, "
                    . 'active: ' . ($category->getIsActive() ? 'Yes' : 'No') . ', '
                    . 'virtual: ' . ($category->getIsVirtualCategory() ? 'Yes' : 'No') . ', '
                    . 'hasChildren: ' . ($category->hasChildren() ? 'Yes' : 'No') . ', '
                    . 'excludedCategories=' . implode(',', $excludedCategories);
                fwrite($handle, $message . PHP_EOL);
                fclose($handle);
            }

Search in that log for a text "category: id=2," and you will see it duplicates in lines: 383, 573, 763 and so on. So the process goes through the same categories again and again.

The issue is the `$excludedCategories` is not passed by reference e.g. `&$excludedCategories` and values added in further recursion levels are missed when you came back to the 1st recursion level.

Another potential issue is in [getStandardCategoriesQuery](https://github.com/Smile-SA/elasticsuite/blob/2.8.3/src/module-elasticsuite-virtual-category/Model/Rule.php#L268) method where the $excludedCategories variable is passed to the [getSearchQuery](https://github.com/Smile-SA/elasticsuite/blob/2.8.3/src/module-elasticsuite-catalog-rule/Model/Rule/Condition/Product/QueryBuilder.php#L77) method which doesn't support it.

### Dirty hotfix

If I replace the usage of `$excludedCategories` to the `static` variable as shown below the issue disappears. But it is a dirty fix, I suppose a better approach should be applied here.

    public function getCategorySearchQuery($category, $excludedCategories = []): ?QueryInterface
    {
        $query         = null;

        if (!is_object($category)) {
            $category = $this->categoryFactory->create()->setStoreId($this->getStoreId())->load($category);
        }

        static $ids = [];

        if (!in_array($category->getId(), $ids)) {
            $excludedCategories[] = $category->getId();
            $ids[] = $category->getId();

            if ((bool) $category->getIsVirtualCategory() && $category->getIsActive()) {
                $query = $this->getVirtualCategoryQuery($category, $excludedCategories, $category->getData('virtual_category_root'));
            } elseif ($category->getId() && $category->getIsActive()) {
                $query = $this->getStandardCategoryQuery($category, $excludedCategories);
            }
            if ($query && $category->hasChildren()) {
                $query = $this->addChildrenQueries($query, $category, $excludedCategories);
            }
        }

        return $query;
    }

**Comment by rbayet** [comment]
Hello @yutv,

Has your "Virtual Category 2" (`1/2/91` I guess in your log file) a "Virtual Category Root" defined and if so, which is it ?
Do you other virtual categories have a virtual category root defined ?

Regards,

**Comment by yutv** [comment]
Hello @rbayet,

Virtual Category 1 (id=38) has Virtual Category Root=Some Category (id=41)
Virtual Category 2 (id=91) and most virtual subcategories have Virtual Category Root=Default Category (id=2). 
Virtual Category 3 (id=303) has Virtual Category Root=Default Category (id=2).

Here is a new log https://pastebin.com/N6i3vFm4 from the following code placed in the line 150. As you see some virtual categories (e.g. id=210, id=298) have virtual category root: NULL, maybe that is the issue.

            if ($handle = fopen('/tmp/test.log', 'a+')) {
                $message = "category: id={$category->getId()}, path={$category->getData('path')}, "
                    . 'active: ' . ($category->getIsActive() ? 'Yes' : 'No') . ', '
                    . 'virtual: ' . ($category->getIsVirtualCategory() ? 'Yes' : 'No') . ', '
                    . 'virtual category root: ' . var_export($category->getData('virtual_category_root'), true) . ', '
                    . 'hasChildren: ' . ($category->hasChildren() ? 'Yes' : 'No') . ', '
                    . 'excludedCategories=' . implode(',', $excludedCategories);
                fwrite($handle, $message . PHP_EOL);
                fclose($handle);
            }

**Comment by PierreLeMaguer** [comment]
Hi @rbayet ,

can you tell us if the following PR is needed ? Or the issue comes from that some virtual categories (e.g. id=210, id=298) have virtual category root: NULL.

Thanks,

**Comment by rbayet** [comment]
Hello @PierreLeMaguer,

No, the issue is the not that some virtual categories have a NULL virtual category root.

The issue is indeed a blind spot related to having several children categories sharing an "ancestor" virtual category root.
As best I can describe it based on when I looked at it before the holidays season, the current code works correctly when browsing a subtree and collecting virtual categories along the way, but following the virtual category root might be in essence a "jump" in the tree.

Long story short, a PR is needed indeed, I'm currently looking at @yutv's one.

Regards,

**Comment by rbayet** [comment]
OK, I have looked at it thoroughly locally, and there is no infinite recursion in the situation described.
That does not mean it could not be achieved :smiling_imp: 

But it is a naïve recursion nonetheless where the whole category tree will be navigated through for each children (virtual) categories of the initial category (because they all point to the same root category).
And navigating through the whole tree means taking into account the sibling virtual categories, etc.
So the complexity is probably something like O(n^2 log n) :disappointed: 

We'll continue looking at @yutv's PR and decide if we want to go this route.

Regards

**Comment by rbayet** [comment]
Original PR would only work without affecting the boolean clauses only in the given example.
The new PR behaves like the merchandiser preview/frontend works, by ignoring the store root category when it is declared as the virtual root category.

This means there might still be an issue under some circumstances where all virtual categories share a common ancestor virtual root category, but the original PR did not preserve excursion into the original virtual category children.
